python/lsst/butler/datastore/posix/datastore.py

```diff
@@ -0,0 +1,233 @@
+        """
+        return yaml.dump(self._data, stream)
+
+    def loadFromFile(self, filename):
+        """Load configuration from a file.
+
+        Parameters
+        ----------
+        filename : `str`
+            The file to load from.
+        """
+        super().loadFromFile(filename)
+
+    def dumpToFile(self, filename):
+        """Dump configuration to a file.
+
+        Parameters
+        ----------
+        filename : `str`
+            The file to dump to.
+        """
+        super().dumpToFile(filename)
+
+    @property
+    def formatters(self):
+        return self._data.formatters
+
+    def _loadDefaultConfiguration(self):
+        self._data.formatters = {
+            "SourceCatalog": "lsst.butler.datastore.posix.fitsCatalogFormatter.FitsCatalogFormatter"}
+
+
+class PosixDatastore(Datastore):
+    """Basic POSIX filesystem backed Datastore.
+    """
+
+    def __init__(self, root="./", create=False, configFile=None):
+        """Construct a Datastore backed by a POSIX filesystem.
+
+        Parameters
+        ----------
+        root : `str`
+            Root directory.
+        create : `bool`
+            Create root directory if it does not exist.
+
+        Raises
+        ------
+        `ValueError` : If root directory does not exist and `create` is `False`.
+        """
+        if not os.path.isdir(root):
+            if not create:
+                raise ValueError("No valid root at: {0}".format(root))
+            safeMakeDir(root)
+        self.root = root
+
+        self.locationFactory = LocationFactory(self.root)
+
+        self.config = PosixDatastoreConfig()
+        if configFile:
+            self.config.loadFromFile(configFile)
+
+        self.formatterFactory = FormatterFactory()
+        for storageClass, formatter in self.config.formatters.items():
+            self.formatterFactory.registerFormatter(storageClass, formatter)
+
+    def get(self, uri, storageClass, parameters=None):
+        """Load an `InMemoryDataset` from the store.
+
+        Parameters
+        ----------
+        uri : `str`
+            a Universal Resource Identifier that specifies the location of the stored `Dataset`.
+        storageClass : `StorageClass`
+            the `StorageClass` associated with the `DatasetType`.
+        parameters : `dict`
+            `StorageClass`-specific parameters that specify a slice of the `Dataset` to be loaded.
+
+        Returns
+        -------
+        inMemoryDataset : `InMemoryDataset`
+            Requested `Dataset` or slice thereof as an `InMemoryDataset`.
+        """
+        formatter = self.formatterFactory.getFormatter(storageClass)
+        location = self.locationFactory.fromUri(uri)
+        if not os.path.exists(location.path):
+            raise ValueError("No such file: {0}".format(location.uri))
+        return formatter.read(FileDescriptor(location, storageClass.type, parameters))
+
+    def put(self, inMemoryDataset, storageClass, path, typeName=None):
```

Comment:
I'm wondering if you need to allow parameters to be passed here as well.  It certainly wouldn't hurt to default it to `None`, just in case we have a data storage that needs to know something for both reads and writes.

I wonder actually if it wouldn't be good for all these access commands to just forward `kwargs` to lower level calls where they could be used or not.

Reply:
Possibly. Although currently the API for `get` and `put` was assumed to be fixed by DMTN-056 with only lower layers to be prototyped. We should likely add parameters/flags as options, although it is a philosophical question if these should be fixed in configuration of a `Datastore` or supplied on `put/get` (e.g. compression settings, file formats, etc.). This can create problems for mismatched `get` and `put` flags.